In [102]:
import pandas as pd
purchase_df = pd.read_csv('data/purchase.csv', sep=',')
events_df = pd.read_csv('data/7_4_Events.csv', sep=',')
condition = (events_df.start_time >= '2018-01-01') & (events_df.start_time < '2019-01-01') & (events_df.event_type =='registration')
registered = events_df[condition]['user_id'].to_list()
events_2018 = events_df[events_df.user_id.isin(registered)]
events_2018.start_time = pd.to_datetime(events_2018.start_time, format='%Y-%m-%dT%H:%M:%S')
purchase_2018 = purchase_df[purchase_df['user_id'].isin(registered)]
purchase_2018.event_datetime = pd.to_datetime(purchase_2018.event_datetime, format='%Y-%m-%dT%H:%M:%S')
purchase_2018["event_type"] = 'purchase'
events_2018 = events_2018.rename(columns={"id": "event_id"})
purchase_2018 = purchase_2018.rename(columns={"id": "purchase_id"})
purchase_2018 = purchase_2018.rename(columns={"event_datetime": "start_time"}) 
total_events_df = pd.concat([events_2018,purchase_2018],sort=False)
total_events_df = total_events_df.reset_index(drop=True).sort_values('start_time')
registration_df = total_events_df[total_events_df['event_type'] == 'registration']
registration_df = registration_df[["user_id", "start_time"]].rename(columns={"start_time": "registration_time"})
level_choice_df = total_events_df[total_events_df["event_type"] == "level_choice"]
level_choice_df = level_choice_df[["user_id", "start_time"]].rename(columns={"start_time": "level_choice_time"})
first_purchase_df = total_events_df[total_events_df['event_type'] == 'purchase']
first_purchase_df = first_purchase_df[["user_id", "start_time"]].rename(columns={"start_time": "first_purchase_time"})
# Создаем группы пользователей в зависимости от выбранной сложности:
choiced_easy_df = total_events_df[total_events_df['selected_level'] == 'easy']
choiced_medium_df = total_events_df[total_events_df['selected_level'] == 'medium']
choiced_hard_df = total_events_df[total_events_df['selected_level'] == 'hard']
# Для каждой группы создаем датафреймы: с событиями выбора уровня сложности, с событиями оплаты и регистрации:
easy_df = total_events_df[total_events_df["selected_level"] == "easy"].sort_values('start_time').drop_duplicates('user_id')
easy_users_ids = easy_df['user_id']
easy_purchase_df = total_events_df[total_events_df["event_type"] == "purchase"][total_events_df["user_id"].isin(easy_users_ids)].sort_values('start_time').drop_duplicates('user_id')
easy_registration_df = total_events_df[total_events_df["event_type"] == "registration"][total_events_df["user_id"].isin(easy_users_ids)].sort_values('start_time').drop_duplicates('user_id')
medium_df = total_events_df[total_events_df["selected_level"] == "medium"].sort_values('start_time').drop_duplicates('user_id')
medium_users_ids = medium_df['user_id']
medium_purchase_df = total_events_df[total_events_df["event_type"] == "purchase"][total_events_df["user_id"].isin(medium_users_ids)].sort_values('start_time').drop_duplicates('user_id')
medium_registration_df = total_events_df[total_events_df["event_type"] == "registration"][total_events_df["user_id"].isin(medium_users_ids)].sort_values('start_time').drop_duplicates('user_id')
hard_df = total_events_df[total_events_df["selected_level"] == "hard"].sort_values('start_time').drop_duplicates('user_id')
hard_users_ids = hard_df['user_id']
hard_purchase_df = total_events_df[total_events_df["event_type"] == "purchase"][total_events_df["user_id"].isin(hard_users_ids)].sort_values('start_time').drop_duplicates('user_id')
hard_registration_df = total_events_df[total_events_df["event_type"] == "registration"][total_events_df["user_id"].isin(hard_users_ids)].sort_values('start_time').drop_duplicates('user_id')
# Для каждой группы подсчитываем процент пользователей, оплативших пакеты вопросов:
percent_choiced_easy = len(easy_purchase_df)/len(easy_df)
print("Процент оплат пользователей, выбравших уровень сложности easy (от общего числа пользователей данной группы): {:.2%}".format(percent_choiced_easy))
percent_choiced_medium = len(medium_purchase_df)/len(medium_df)
print("Процент оплат пользователей, выбравших уровень сложности medium (от общего числа пользователей данной группы): {:.2%}".format(percent_choiced_medium))
percent_choiced_hard = len(hard_purchase_df)/len(hard_df)
print("Процент оплат пользователей, выбравших уровень сложности hard (от общего числа пользователей данной группы): {:.2%}".format(percent_choiced_hard))

Процент оплат пользователей, выбравших уровень сложности easy (от общего числа пользователей данной группы): 7.72%
Процент оплат пользователей, выбравших уровень сложности medium (от общего числа пользователей данной группы): 20.86%
Процент оплат пользователей, выбравших уровень сложности hard (от общего числа пользователей данной группы): 35.39%


/var/folders/2t/41ymw9416szdry3f82rfxvr80000gn/T/ipykernel_18409/3430041532.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  events_2018.start_time = pd.to_datetime(events_2018.start_time, format='%Y-%m-%dT%H:%M:%S')
/var/folders/2t/41ymw9416szdry3f82rfxvr80000gn/T/ipykernel_18409/3430041532.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  purchase_2018.event_datetime = pd.to_datetime(purchase_2018.event_datetime, format='%Y-%m-%dT%H:%M:%S')
/var/folders/2t/41ymw9416szdry3f82rfxvr80000gn/T/ipykernel_1

Вывод 1. Есть прямая зависимость между выбранным уровнем сложности и вероятностью оплаты у пользователей. Так наивысший процент оплат у пользователей, выбравших уровень сложности hard, он составляет 35.39%, а наименьший процент у пользователей, выбравших уровень сложности easy, он состовляет 7.72%.

In [107]:

# Для каждой группы подсчитываем среднее время между событиями оплаты и выбором уровня сложности:
easy_df1 = easy_df[["user_id", "start_time"]].rename(columns={"start_time": "level_select_time"})
easy_purchase_df1 = easy_purchase_df[["user_id", "start_time"]].rename(columns={"start_time": "purchase_time"})
df2 = easy_df1.merge(easy_purchase_df1, how="inner", on="user_id")
df2["timedelta"] = df2["purchase_time"] - df2["level_select_time"]
print(df2["timedelta"].mean())
medium_df1 = medium_df[["user_id", "start_time"]].rename(columns={"start_time": "level_select_time"})
medium_purchase_df1 = medium_purchase_df[["user_id", "start_time"]].rename(columns={"start_time": "purchase_time"})
df1 = medium_df1.merge(medium_purchase_df1, how="inner", on="user_id")
df1["timedelta"] = df1["purchase_time"] - df1["level_select_time"]
print(df1["timedelta"].mean())
hard_df1 = hard_df[["user_id", "start_time"]].rename(columns={"start_time": "level_select_time"})
hard_purchase_df1 = hard_purchase_df[["user_id", "start_time"]].rename(columns={"start_time": "purchase_time"})
df3 = hard_df1.merge(hard_purchase_df1, how="inner", on="user_id")
df3["timedelta"] = df3["purchase_time"] - df3["level_select_time"]
print(df3["timedelta"].mean())

3 days 14:58:52.941798941
3 days 23:14:13.165118679
3 days 07:20:41.420814479


Вывод 2. Минимальное среднее время между событиями оплаты и выбором уровня сложности у пользователей, выбравших уровень сложности hard. Оно составляет 3 дня 07:20:41, а максимальное среднее время между событиями оплаты и выбором уровня сложности у пользователей, выбравших уровень сложности medium. Оно составляет 3 дня 23:14:13.

In [108]:

# Проверяем, существует ли разница во времени между событиями регистрации и оплаты для каждой из групп:
merged_df_1 = easy_registration_df.merge(easy_purchase_df1, on="user_id", how="inner")
merged_df_1["timedelta"] = (merged_df_1["purchase_time"] - merged_df_1["start_time"])
print(merged_df_1["timedelta"].describe())
merged_df_2 = medium_registration_df.merge(medium_purchase_df1, on="user_id", how="inner")
merged_df_2["timedelta"] = (merged_df_2["purchase_time"] - merged_df_2["start_time"])
print(merged_df_2["timedelta"].describe())
merged_df_3 = hard_registration_df.merge(hard_purchase_df1, on="user_id", how="inner")
merged_df_3["timedelta"] = (merged_df_3["purchase_time"] - merged_df_3["start_time"])
print(merged_df_3["timedelta"].describe())

count                          189
mean     3 days 22:10:23.211640211
std      2 days 07:14:41.062010764
min                0 days 04:36:58
25%                2 days 01:12:12
50%                3 days 11:00:23
75%                5 days 10:24:59
max               11 days 00:35:04
Name: timedelta, dtype: object
count                          969
mean     4 days 06:12:06.576883384
std      2 days 06:25:57.480868026
min                0 days 08:39:24
25%                2 days 08:46:51
50%                4 days 03:35:26
75%                5 days 23:51:27
max               10 days 20:34:02
Name: timedelta, dtype: object
count                          442
mean     3 days 14:55:19.257918552
std      1 days 22:22:52.441896774
min                0 days 09:41:39
25%         1 days 23:36:25.500000
50%         3 days 10:10:04.500000
75%         5 days 03:30:07.750000
max                8 days 14:21:29
Name: timedelta, dtype: object


Вывод 3. Существует разница во времени между событиями регистрации и оплаты. Минимальное среднее время между событиями оплаты и регистрацией у пользователей, выбравших уровень сложности hard. Оно составляет 3 days 14:55:19.257918552, а максимальное среднее время между событиями оплаты и регистрацией у пользователей, выбравших уровень сложности medium. Оно составляет 4 дня 06:12:06.